### PR TITLE
Allow zero-length input in rnp_input_from_memory()

### LINKED ISSUE #2173: Encryption and decryption of zero length message raise error
## Description

I encountered an error when trying to encrypt and decrypt a zero length message using the rnp library. 
The steps to reproduce the error are as follows

## Steps to Reproduce

1. Run the generate.c example from https://github.com/rnpgp/rnp/blob/main/src/examples/generate.c
2. Run the encrypt.c example from https://github.com/rnpgp/rnp/blob/main/src/examples/encrypt.c

Ensure  ("Encryption succeeded. Encrypted message written to file encrypted.asc")

1. Change line 42 of encrypt.c from `const char *     message = "RNP encryption sample message";` to `const char *     message = "";`
2. Run the encrypt.c example again and see the error message: `failed to create input object`
3. Create an empty file with `touch myEmptyFile`
4. Encrypt the empty file with rnp or gnupg
5. Decrypt the encrypted file to memory and see the error message: `failed to create input object`

## Expected Behavior

I expected the encryption and decryption of a zero length message to succeed without any error, or at least to fail gracefully with a meaningful error message.

## Actual Behavior

The encryption and decryption of a zero length message failed with a cryptic error message: `failed to create input object`

### More info
Notice that the error happens when encrypt from buffer, ot decrypt to buffer. 
File to file, is working fine.

### PR DESCRIPTION
**Pull request by desvxx**
fixes #2173

### CODE REVIEW

File: src/lib/rnp.cpp [hunk 1]
```diff
@@ -2099,9 +2099,6 @@ try {
     if (!input || !buf) {
```
Comment (ni4) [inline src/lib/rnp.cpp:2099]: Please also allow to work with NULL input buffer and zero input size, and add the corresponding test.

File: src/tests/ffi-enc.cpp [hunk 2]
```diff
@@ -381,7 +381,7 @@ TEST_F(rnp_tests, test_ffi_encrypt_set_cipher)
     assert_rnp_failure(
       rnp_input_from_memory(NULL, (const uint8_t *) plaintext, strlen(plaintext), false));
     assert_rnp_failure(rnp_input_from_memory(&input, NULL, strlen(plaintext), false));
-    assert_rnp_failure(rnp_input_from_memory(&input, (const uint8_t *) plaintext, 0, false));
+    assert_rnp_success(rnp_input_from_memory(&input, (const uint8_t *) plaintext, 0, false));
```
Comment (ni4) [inline src/tests/ffi-enc.cpp:384]: Please also make sure that encrypting/signing of zero-size input works as well.

**Review by ni4 — CHANGES_REQUESTED** [review]
Please see the comments.

File: src/lib/rnp.cpp [hunk 3]
```diff
@@ -2096,11 +2096,12 @@ FFI_GUARD
 rnp_result_t
 rnp_input_from_memory(rnp_input_t *input, const uint8_t buf[], size_t buf_len, bool do_copy)
 try {
-    if (!input || !buf) {
+    if (!input || (!buf && buf_len > 0)) {
```
Comment (ni4) [inline src/lib/rnp.cpp:2099]: It would be logical to replicate behaviour of `init_mem_src()` here, which allows null buffer for the zero-size input.

File: src/tests/ffi-enc.cpp [hunk 4]
```diff
@@ -370,20 +370,29 @@ TEST_F(rnp_tests, test_ffi_encrypt_pass_provider)
     rnp_ffi_destroy(ffi);
 }
 
-TEST_F(rnp_tests, test_ffi_encrypt_set_cipher)
+class ffi_encrypt_tests : public rnp_tests, public testing::WithParamInterface<const char *> {
+};
+
+TEST_P(ffi_encrypt_tests, test_ffi_encrypt_set_cipher)
```
Comment (ni4) [inline src/tests/ffi-enc.cpp:376]: Could you please create a new test case instead of heavily modifying this one?

File: src/tests/ffi-enc.cpp [hunk 5]
```diff
@@ -370,20 +370,29 @@ TEST_F(rnp_tests, test_ffi_encrypt_pass_provider)
...
+    const char * plaintext = GetParam();
+    const size_t plaintext_size = plaintext ? strlen(plaintext) : 0;
+    const char * decrypted =
+      plaintext ? plaintext : ""; // NULL input gives us "" when decrypted
     assert_rnp_failure(
-      rnp_input_from_memory(NULL, (const uint8_t *) plaintext, strlen(plaintext), false));
-    assert_rnp_failure(rnp_input_from_memory(&input, NULL, strlen(plaintext), false));
-    assert_rnp_failure(rnp_input_from_memory(&input, (const uint8_t *) plaintext, 0, false));
+      rnp_input_from_memory(NULL, (const uint8_t *) plaintext, plaintext_size, false));
+    if (plaintext == NULL) { // Corner case
```
Comment (ni4) [inline src/tests/ffi-enc.cpp:389]: We prefer `if (!plaintext)` notation in such cases.

File: src/tests/ffi-enc.cpp [hunk 6]
```diff
@@ -370,20 +370,29 @@ TEST_F(rnp_tests, test_ffi_encrypt_pass_provider)
...
+        assert_rnp_failure(rnp_input_from_memory(&input, NULL, 5, false));
+    }
+    assert_rnp_success(rnp_input_from_memory(&input, (const uint8_t *) plaintext, 0, false));
```
Comment (ni4) [inline src/tests/ffi-enc.cpp:392]: Will not this interfer with the previous code line as plaintext is null?

**Review by ni4 — CHANGES_REQUESTED** [review]
Please see the review comments.

**Review by ni4 — APPROVED** [review]
LGTM, thanks!

**Review by maxirmx — APPROVED** [review]
OK
